app/src/main/java/com/github/ravifrancesco/softwaredevproject/ElevationMap.java

```diff
@@ -0,0 +1,200 @@
+        try {
+            HttpClient httpClient = HttpClientBuilder.create().build();
+            HttpRequestBase base = new HttpGet(url.toString());
+            HttpResponse response = httpClient.execute(base);
+
+            if (response.getStatusLine().getStatusCode() == 200) {
+                parseResponse(response);
+            } else {
+                Log.d("3d MAP", "Http error code: " + response.getStatusLine().getStatusCode());
+            }
+
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * This method handles the parsing of the response obtained via HTTP request. It saves the
+     * size of the matrix to build and passes them to the buildMapGrid method.
+     *
+     * @param response  HTTPResponse to parse.
+     */
+    private void parseResponse(HttpResponse response) {
+
+        try {
+            Scanner responseObj = new Scanner(response.getEntity().getContent());
+            int nCol =  Integer.parseInt(responseObj.nextLine().replaceAll("[\\D]", ""));
+            int nRow = Integer.parseInt(responseObj.nextLine().replaceAll("[\\D]", ""));
+            buildMapGrid(nRow, nCol, responseObj);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    /**
+     * This method builds the matrix and saves it in the object.
+     *
+     * @param nRow          height of the matrix.
+     * @param nCol          width of the matrix.
+     * @param responseObj   Scanner to read the HTTPResponse.
+     */
+    private void buildMapGrid(int nRow, int nCol, Scanner responseObj) {
+
+        // skip lines to get to first row
+        responseObj.nextLine();
+        responseObj.nextLine();
+        responseObj.nextLine();
+        responseObj.nextLine();
+
+        // build matrix
+        this.topographyMap = IntStream.range(0, nRow)
+                .boxed()
+                .map(i -> Arrays.stream(responseObj.nextLine().substring(1).split(" ", nCol))
+                        .mapToInt(Integer::parseInt)
+                        .toArray())
+                .toArray(int[][]::new);
+
+        Log.d("3D MAP", "Generated Map with size (" + nRow + ", " + nCol +")");
+
+    }
+
+    /**
+     * Public method to retrieve the elevation map.
+     *
+     * @return  an int[][] array representing the elevation map.
+     */
+    public int[][] getTopographyMap() {
+
+        updateBoundingBox();
+        return topographyMap;
+
+    }
+
+    /**
+     * This method handles the updating of the elevation map. First the distance between the center
+     * of the current bounding box and the current user location is computed.
+     *
+     * If less than MINIMUM_DISTANCE_FOR_UPDATE no new map is downloaded, otherwise a new map is
+     * downloaded.
+     */
+    private void updateBoundingBox() {
+
+        Log.d("3D MAP",  "Updating");
+        POIPoint oldBoundingCenter = new POIPoint(boundingBox.getCenterWithDateLine());
```

Comment:
I am a bit unsure on how often this method is called. I therefore wonder if it would save a bit of computing if we create the boundingCenter and save it when when we set the bounding box. So we only need to create an POIPoint one time instead of every time updateBoundingBox is called.

Reply:
Good point, I'll change it